ocf_datapipes/load/gsp/database.py

```diff
@@ -0,0 +1,195 @@
+""" Function to get data from live database """
+import logging
+import os
+from datetime import datetime, timedelta, timezone
+from typing import List
+
+import numpy as np
+import pandas as pd
+import xarray as xr
+from nowcasting_datamodel import N_GSP
+from nowcasting_datamodel.connection import DatabaseConnection
+from nowcasting_datamodel.models.base import Base_Forecast
+from nowcasting_datamodel.models.gsp import GSPYield, GSPYieldSQL, Location
+from nowcasting_datamodel.read.read_gsp import get_gsp_yield
+from torchdata.datapipes import functional_datapipe
+from torchdata.datapipes.iter import IterDataPipe
+
+from ocf_datapipes.load.gsp.utils import put_gsp_data_into_an_xr_dataarray
+from ocf_datapipes.utils.eso import get_gsp_shape_from_eso
+
+logger = logging.getLogger(__name__)
+
+
+@functional_datapipe("open_gsp_from_database")
+class OpenGSPFromDatabaseIterDataPipe(IterDataPipe):
+    """Get and open the GSP data"""
+
+    def __init__(
+        self,
+        history_minutes: int = 90,
+        interpolate_minutes: int = 60,
+        load_extra_minutes: int = 60,
+    ):
+        """
+        Get and open the GSP data
+
+        Args:
+            history_minutes: How many history minutes to use
+            interpolate_minutes:  How many minutes to interpolate
+            load_extra_minutes: How many extra minutes to load
+        """
+
+        self.interpolate_minutes = interpolate_minutes
+        self.load_extra_minutes = load_extra_minutes
+        self.history_duration = timedelta(minutes=history_minutes)
+
+    def __iter__(self) -> xr.DataArray:
+        """Get and return GSP data"""
+
+        logger.debug("Getting GSP data")
+
+        gsp_pv_power_mw_df, gsp_capacity = get_gsp_power_from_database(
+            history_duration=self.history_duration,
+            interpolate_minutes=self.interpolate_minutes,
+            load_extra_minutes=self.load_extra_minutes,
+        )
+
+        # get shape file
+        gsp_id_to_shape = get_gsp_shape_from_eso(return_filename=False)
+
+        # Ensure the centroids have the same GSP ID index as the GSP PV power:
+        gsp_id_to_shape = gsp_id_to_shape.loc[gsp_pv_power_mw_df.columns]
+
+        data_array = put_gsp_data_into_an_xr_dataarray(
+            gsp_pv_power_mw=gsp_pv_power_mw_df.astype(np.float32),
+            time_utc=gsp_pv_power_mw_df.index.values,
+            gsp_id=gsp_pv_power_mw_df.columns,
+            # TODO: Try using `gsp_id_to_shape.geometry.envelope.centroid`. See issue #76.
+            x_osgb=gsp_id_to_shape.geometry.centroid.x.astype(np.float32),
+            y_osgb=gsp_id_to_shape.geometry.centroid.y.astype(np.float32),
+            capacity_megawatt_power=gsp_capacity.astype(np.float32),
+        )
+
+        del gsp_id_to_shape, gsp_pv_power_mw_df
+        while True:
+            logger.debug(data_array)
```

Comment:
Would we want to just log this once right before the True? Then it wouldn't clog up debug logs by showing up lots of times?

Reply:
thanks, yea ill remove that